Topic/anmajumdar/five tuple dns

anubhabMajumdar opened the pull request:
# feat(dns-gadget): Add more fields to DNS gadget

Add the following fields to `DNS` gadget event:
-  SrcIP
- DstIP
- SrcPort
- DstPort
- Protocol

## How to use

```bash
$ ./kubectl-gadget trace dns -n default                                                                                                                                                                                                                                                                                                            ok | 6s | 17:26:02
NODE                                                        NAMESPACE                                                   POD                                                         PID        TID        COMM             SRCIP                          DSTIP                          SRCP… DSTP… PROTO QR TYPE      QTYPE      NAME                                                       RCODE                          NUMANSW…
minikube-docker                                             default                                                     alpine                                                      2661242    2661242    nslookup         10.244.0.5                     10.96.0.10                     51025 53    udp   Q  OUTGOING  A          bing.com.                                                                                 0
minikube-docker                                                                                                                                                                     0          0                           10.244.0.5                     10.96.0.10                     51025 53    udp   Q  OTHERHOST A          bing.com.                                                                                 0
minikube-docker                                                                                                                                                                     0          0                           10.244.0.5                     10.244.0.2                     51025 53    udp   Q  OUTGOING  A          bing.com.                                                                                 0
minikube-docker                                                                                                                                                                     0          0                           10.244.0.2                     192.168.65.2                   46205 53    udp   Q  OTHERHOST A          bing.com.                                                                                 0
minikube-docker                                                                                                                                                                     0          0                           10.244.0.2                     127.0.0.11                     46205 40648 udp   Q  HOST      A          bing.com.                                                                                 0
minikube-docker                                                                                                                                                                     2582       2703       dockerd          192.168.49.2                   192.168.65.7                   54400 53    udp   Q  OUTGOING  A          bing.com.                                                                                 0
minikube-docker                                                                                                                                                                     2582       2703       dockerd          192.168.65.7                   192.168.49.2                   53    54400 udp   R  HOST      A          bing.com.                                                  NoError                        2
minikube-docker                                                                                                                                                                     0          0                           192.168.65.2                   10.244.0.2                     53    46205 udp   R  OUTGOING  A          bing.com.                                                  NoError                        2
minikube-docker                                                                                                                                                                     0          0                           192.168.65.2                   10.244.0.2                     53    46205 udp   R  OUTGOING  A          bing.com.                                                  NoError                        2
minikube-docker                                                                                                                                                                     0          0                           10.244.0.2                     10.244.0.5                     53    51025 udp   R  OTHERHOST A          bing.com.                                                  NoError                        2
minikube-docker                                                                                                                                                                     0          0                           10.96.0.10                     10.244.0.5                     53    51025 udp   R  OUTGOING  A          bing.com.                                                  NoError                        2
minikube-docker                                             default                                                     alpine                                                      2661242    2661242    nslookup         10.96.0.10                     10.244.0.5                     53    51025 udp   R  HOST      A          bing.com.                                                  NoError                        2
minikube-docker                                             default                                                     alpine                                                      2661601    2661601    nslookup         10.244.0.5                     10.96.0.10                     46024 53    udp   Q  OUTGOING  AAAA       google.com.                                                                               0
minikube-docker                                                                                                                                                                     0          0                           10.244.0.5                     10.96.0.10                     46024 53    udp   Q  OTHERHOST AAAA       google.com.                                                                               0
minikube-docker                                                                                                                                                                     0          0                           10.244.0.5                     10.244.0.2                     46024 53    udp   Q  OUTGOING  AAAA       google.com.                                                                               0
minikube-docker                                                                                                                                                                     0          0                           10.244.0.2                     192.168.65.2                   38867 53    udp   Q  OTHERHOST AAAA       google.com.                                                                               0
minikube-docker                                                                                                                                                                     2582       1959886    dockerd          10.244.0.2                     127.0.0.11                     38867 40648 udp   Q  HOST      AAAA       google.com.                                                                               0
minikube-docker                                                                                                                                                                     2582       1959886    dockerd          192.168.49.2                   192.168.65.7                   43901 53    udp   Q  OUTGOING  AAAA       google.com.                                                                               0
minikube-docker                                                                                                                                                                     2582       1959886    dockerd          192.168.65.7                   192.168.49.2                   53    43901 udp   R  HOST      AAAA       google.com.                                                NoError                        1
minikube-docker                                                                                                                                                                     0          0                           192.168.65.2                   10.244.0.2                     53    38867 udp   R  OUTGOING  AAAA       google.com.                                                NoError                        1
minikube-docker                                                                                                                                                                     0          0                           192.168.65.2                   10.244.0.2                     53    38867 udp   R  OUTGOING  AAAA       google.com.                                                NoError                        1
minikube-docker                                                                                                                                                                     0          0                           10.244.0.2                     10.244.0.5                     53    46024 udp   R  OTHERHOST AAAA       google.com.                                                NoError                        1
minikube-docker                                                                                                                                                                     0          0                           10.96.0.10                     10.244.0.5                     53    46024 udp   R  OUTGOING  AAAA       google.com.                                                NoError                        1
minikube-docker                                             default                                                     alpine                                                      2661601    2661601    nslookup         10.96.0.10                     10.244.0.5                     53    46024 udp   R  HOST      AAAA       google.com.                                                NoError                        1
```


## Testing done

- Unit tests are passing
- Modified the DNS integration test to test for the new fields
- Manually verified in a `minikube` cluster, snippet added above

anubhabMajumdar commented on integration/inspektor-gadget/trace_dns_test.go:
```diff
@@ -127,20 +156,37 @@ func TestTraceDns(t *testing.T) {
-			return ExpectEntriesToMatch(output, normalize, expectedEntries...)
-		},
-	}
+			compFn := func(a, b any) bool {
+				x := a.(*tracednsTypes.Event)
+				y := b.(*tracednsTypes.Event)
+				return x.Comm == y.Comm &&
+					x.Qr == y.Qr &&
+					x.Nameserver == y.Nameserver &&
+					x.PktType == y.PktType &&
+					x.DNSName == y.DNSName &&
+					x.QType == y.QType &&
+					x.Rcode == y.Rcode &&
+					x.Latency == y.Latency &&
+					x.NumAnswers == y.NumAnswers &&
+					reflect.DeepEqual(x.Addresses, y.Addresses) &&
+					x.Uid == y.Uid &&
+					x.Gid == y.Gid &&
+					x.Protocol == y.Protocol &&
+					((x.Qr == tracednsTypes.DNSPktTypeQuery &&
+						x.DstPort == y.DstPort && x.SrcIP == y.SrcIP) ||
+						(x.Qr == tracednsTypes.DNSPktTypeResponse &&
+							x.SrcPort == y.SrcPort &&
+							x.DstIP == y.DstIP))
+			}
```
Under the default approach, we are matching all the fields generated by gadget event. I think for the DNS gadget event, we are interested in a subset of the fields (like the destination port for query should be 53) and not necessarily all fields (we don't care so much about the source port of the bustbox pod running the `nslookup`). Maybe this could be beneficial for other gadget tests going forward? We maybe interested to test only the fields we can control during test and test them?

blanquicet replied:
I agree with Francis, the `normalize` function should do the trick you're looking for.

anubhabMajumdar commented on integration/helpers.go:
```diff
@@ -121,7 +114,7 @@ func ExpectAllInMultipleArrayToMatch[T any](t *testing.T, output string, normali
 }
 
-func expectEntriesToMatch[T any](fn comparisonFn, entries []*T, expectedEntries ...*T) error {
+func expectEntriesToMatch[T any](t *testing.T, fn comparisonFn, entries []*T, expectedEntries ...*T) {
```
I believe I have rebased on top of this commit? Here's the function definition from `main`
```
func expectEntriesToMatch[T any](t *testing.T, entries []*T, expectedEntries ...*T) 
```

anubhabMajumdar commented on pkg/gadgets/helpers.go:
```diff
@@ -305,3 +319,16 @@ func LoadeBPFSpec(
 	return nil
 }
+
+func ProtoString(proto int) string {
+	// proto definitions:
+	// https://www.iana.org/assignments/protocol-numbers/protocol-numbers.xhtml
+	protoStr := fmt.Sprintf("UNKNOWN#%d", proto)
+	switch proto {
+	case 6:
+		protoStr = "tcp"
+	case 17:
+		protoStr = "udp"
+	}
+	return protoStr
+}
```
Can you please clarify the ` kernel source defining this magics`? Do you mean how we are parsing the protocol from the `__skb` struct?

anubhabMajumdar commented on pkg/gadgets/helpers.go:
```diff
@@ -112,10 +112,24 @@ func Htonl(hl uint32) uint32 {
+	var hl [2]byte
+	determineEndian().PutUint16(hl[:], nl)
+	return *(*uint16)(unsafe.Pointer(&hl[0]))
+}
```
Fixed this!

anubhabMajumdar commented:
> Hi.
> 
> Thank you for working on this. I have some initials comments, particularly regarding integration tests. I will test the gadget itself later today and if everything is OK, I will run the CI.
> 
> Do not hesitate if you have any questions or point you do not understand.
> 
> Best regards.

Thanks for the review, addressed your comments. Let me know if the gadget works for you!

blanquicet commented on integration/inspektor-gadget/trace_dns_test.go:
```diff
@@ -80,9 +100,12 @@ func TestTraceDns(t *testing.T) {
 					Nameserver: dnsServer,
 					PktType:    "OUTGOING",
 					DNSName:    "fake.test.com.",
-					QType:      "A",
+					QType:      "AAAA",
```
This is a mistake currently present in the tests. Thanks for catching it.

anubhabMajumdar replied:
+1

blanquicet commented on pkg/gadgets/trace/dns/types/dns.go:
```diff
@@ -42,6 +42,12 @@ type Event struct {
 	Uid uint32 `json:"uid" column:"uid,template:uid,hide"`
 	Gid uint32 `json:"gid" column:"gid,template:gid,hide"`
 
+	SrcIP    string `json:"srcIP,omitempty" column:"srcIP,template:ipaddr"`
+	DstIP    string `json:"dstIP,omitempty" column:"dstIP,template:ipaddr"`
+	SrcPort  uint16 `json:"srcPort,omitempty" column:"srcPort,template:ipport"`
+	DstPort  uint16 `json:"dstPort,omitempty" column:"dstPort,template:ipport"`
+	Protocol string `json:"protocol,omitempty" column:"proto,maxWidth:5"`
```
IMO, they should be hidden by default. It's too much info to be displayed by default.

anubhabMajumdar replied:
Made these 5 columns `visible=false`.

blanquicet commented on integration/inspektor-gadget/trace_dns_test.go:
```diff
@@ -121,20 +147,36 @@ func TestTraceDns(t *testing.T) {
-	commands := []*Command{
+	// Start the trace gadget and verify the output.
+	commands = []*Command{
 		CreateTestNamespaceCommand(ns),
```
At the beginning of the test, we should create the namespace and then add a cleanup function to delete the ns:

```go
	RunTestSteps([]*Command{CreateTestNamespaceCommand(ns)}, t, WithCbBeforeCleanup(PrintLogsFn(ns)))

	t.Cleanup(func() {
		RunTestSteps([]*Command{DeleteTestNamespaceCommand(ns)}, t, WithCbBeforeCleanup(PrintLogsFn(ns)))
	})
```

Next, we can create the `dnstester`, and then the `busybox`. Otherwise, if the creation of the pods fails for any reason, we won't delete the namespace.

anubhabMajumdar replied:
Done!

blanquicet reviewed (commented):
Thanks @anubhabMajumdar for this contribution! I have a couple of comments but it already looks good.

anubhabMajumdar commented on pkg/gadgets/helpers.go:
```diff
@@ -305,3 +305,16 @@ func LoadeBPFSpec(
+	protoStr := fmt.Sprintf("UNKNOWN#%d", proto)
+	switch proto {
+	case 6:
+		protoStr = "tcp"
+	case 17:
+		protoStr = "udp"
```
+1

anubhabMajumdar commented:
> Thanks @anubhabMajumdar for this contribution! I have a couple of comments but it already looks good.

@eiffel-fl @blanquicet 
Thanks for all the feedback and suggestions! I have addressed all of them. Let me know if this looks OK!

burak-ok reviewed (commented):
Hi,

as the others already said the changes look good, thanks for that :)

Another improvement could be using the `KubeIPResolver` so that IP addresses will be automatically resolved to pod names or service names.
I experimented with it in https://github.com/burak-ok/inspektor-gadget/commit/315746ccf2243adbd852ecca073b8964473f8abf . Feel free to cherry-pick and modify it.

```bash
$ ./kubectl-gadget trace dns -n default -o columns=node,namespace,pod,name,qr,qtype,src
NODE                  NAMESPACE             POD                   NAME                  QR QTYPE      SRC                        
minikube-docker       default               test-pod              google.de.            Q  A          p/default/test-pod:56088   
minikube-docker       default               test-pod              google.de.            Q  AAAA       p/default/test-pod:56088   
minikube-docker       default               test-pod              google.de.            R  AAAA       s/kube-system/kube-dns:53  
minikube-docker       default               test-pod              google.de.            R  A          s/kube-system/kube-dns:53
```

anubhabMajumdar commented on pkg/gadgets/trace/dns/types/dns.go:
```diff
@@ -69,6 +69,13 @@ func GetColumns() *columns.Columns[Event] {
 		col.Visible = false
 	}
 
+	optional := []string{"srcIP", "dstIP", "srcPort", "dstPort", "protocol"}
+	for _, name := range optional {
+		if col, ok := cols.GetColumn(name); ok {
+			col.Visible = false
+		}
+	}
+
```
+1

anubhabMajumdar commented:
> It looks good, you just need to polish the git history. I will run the CI but I expect it to pass.

A couple of CI pipelines failed due to small typo in the test and formatting issues. I have fixed those.

blanquicet commented:
> Another improvement could be using the `KubeIPResolver` so that IP addresses will be automatically resolved to pod names or service names. I experimented with it in [burak-ok@315746c](https://github.com/burak-ok/inspektor-gadget/commit/315746ccf2243adbd852ecca073b8964473f8abf) . Feel free to cherry-pick and modify it.
>
Nice one! Yeah, we should definitely do it. However, I think it could be done in another PR as this one, IMO, is ready to go!

alban reviewed (commented):
Thanks for the PR. I am adding my comments (some are redundant with previous comments).

anubhabMajumdar commented:
> > Another improvement could be using the `KubeIPResolver` so that IP addresses will be automatically resolved to pod names or service names. I experimented with it in [burak-ok@315746c](https://github.com/burak-ok/inspektor-gadget/commit/315746ccf2243adbd852ecca073b8964473f8abf) . Feel free to cherry-pick and modify it.
> 
> Nice one! Yeah, we should definitely do it. However, I think it could be done in another PR as this one, IMO, is ready to go!

Hey, addressed Alban's comments and fixed one more issue with the integration test. Hopefully it will pass this time!

anubhabMajumdar commented:
> Can you please rebase and clean the whole history? It will make it easier to review. I will run the CI again meanwhile.

Pipelines have passed, can I squash and merge from Github?

anubhabMajumdar commented:
> > > > Can you please rebase and clean the whole history? It will make it easier to review. I will run the CI again meanwhile.
> > > 
> > > 
> > > Pipelines have passed, can I squash and merge from Github?
> > 
> > 
> > What do you mean by squash? If you mean squashing all your commits into one, then I do not agree, one PR is not composed of one commit but of several which represent each one semantic changes (adding the features, adding the test, adding the documentation, etc.).
> > So, reworking your history would be welcomed. Once done, I will review it again, potentially give my approval and I will merge it. I would nonetheless like to have another green review before merging.
> 
> As the only modifications you bring to the tests are modifications meant for it to pass, it can be part of the same commit than the one adding the feature. Otherwise, we would end in a red and green commit.
> 
> Nonetheless, we will not use GitHub feature to squash everything before we merge.

Cleaned up the commits; let me know if this works! Thanks!
